src/index.js

```diff
@@ -308,59 +310,99 @@ class JanusAdapter {
     this.session.receive(JSON.parse(event.data));
   }
 
-  async addOccupant(occupantId) {
-    if (this.occupants[occupantId]) {
-      this.removeOccupant(occupantId);
+  addAvailableOccupant(occupantId) {
+    if (this.availableOccupants.indexOf(occupantId) === -1) {
+      this.availableOccupants.push(occupantId);
     }
+  }
 
-    this.leftOccupants.delete(occupantId);
+  removeAvailableOccupant(occupantId) {
+    const idx = this.availableOccupants.indexOf(occupantId);
+    if (idx !== -1) {
+      this.availableOccupants.splice(idx, 1);
+    }
+  }
 
-    var subscriber = await this.createSubscriber(occupantId);
+  syncOccupants(requestedOccupants) {
+    if (requestedOccupants) {
+      this.requestedOccupants = requestedOccupants;
+    }
 
-    if (!subscriber) return;
+    if (!this.requestedOccupants) {
+      return;
+    }
 
-    this.occupants[occupantId] = subscriber;
+    // Add any requested, available, and non-pending occupants.
+    for (let i = 0; i < this.requestedOccupants.length; i++) {
+      const occupantId = this.requestedOccupants[i];
+      if (!this.occupants[occupantId] && this.availableOccupants.indexOf(occupantId) !== -1 && !this.pendingOccupants.has(occupantId)) {
+        this.addOccupant(occupantId);
+      }
+    }
 
-    this.setMediaStream(occupantId, subscriber.mediaStream);
+    // Remove any unrequested and currently added occupants.
+    for (let j = 0; j < this.availableOccupants.length; j++) {
+      const occupantId = this.availableOccupants[j];
+      if (this.occupants[occupantId] && this.requestedOccupants.indexOf(occupantId) === -1) {
+        this.removeOccupant(occupantId);
+      }
+    }
 
-    // Call the Networked AFrame callbacks for the new occupant.
-    this.onOccupantConnected(occupantId);
+    // Call the Networked AFrame callbacks for the updated occupants list.
     this.onOccupantsChanged(this.occupants);
+  }
 
-    return subscriber;
+  async addOccupant(occupantId) {
+    this.pendingOccupants.add(occupantId);
+   this.createSubscriber(occupantId).then(subscriber => {
```

Comment:
eventually we should update `createSubscriber` to encapsulate a retry/backoff loop.